### PR TITLE
Replace Guava caches

### PR DESCRIPTION
**Pull request by randomanderson**
This pull request removes the last Guava dependency for non-testing code 🎉 

### For weak caches:
**Java 7:** Use a combination of `WeakLockFree` and `ConcurrentLinkedHashMap` (CLHM was written by the same guy that went on to write Caffeine and is the recommended solution for < jdk8).  Entries are expunged when `put` or `computeIfAbsent` might hit capacity
**Java 8+:** Use Caffeine

### For other FIFO caches
Use `ConcurrentLinkedHashMap`.  I looked at our `DDCache` implementations and they don't support this use case.

Additionally, I did a small cleanup on the `WeakCache` and `WeakMap` interfaces

For reference, the maps/caches in use after this PR:
|        | Unbounded | Bounded (random removal) | Bounded (FIFO) 
| ---|---|---|---
| Weak Keys | `WeakLockFree` or `synchronized java.util.WeakHashMap` | -- | WeakLockFree+CLHM (jdk7) / Caffeine (jdk8+)
| Strong Keys | `datadog.trace.api.CHMCache` | `datadog.trace.api.FixedSizeCache` | `ConcurrentLinkedHashMap`

### CODE REVIEW

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/AgentTooling.java [hunk 1]
```diff
@@ -27,27 +27,38 @@ static void registerWeakMapProvider() {
...
-      return provider;
+
+      return providerClass.getDeclaredConstructor().newInstance();
+    } catch (ReflectiveOperationException e) {
+      throw new IllegalStateException("Can't load implementation of WeakCache.Provider", e);
     }
-    throw new IllegalStateException("Can't load implementation of WeakCache.Provider");
   }
 
+  private static final long DEFAULT_CACHE_CAPACITY = 32;
```
Comment (randomanderson) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/AgentTooling.java:54]: There is no basis for this number.  The alternative is removing `newWeakCache()` and forcing callers to specify a capacity.

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/bytebuddy/DDCachingPoolStrategy.java [hunk 2]
```diff
@@ -56,23 +59,16 @@
...
 
   /**
    * Single shared Type.Resolution cache -- uses a composite key -- conceptually of loader & name
    */
-  final Cache<TypeCacheKey, TypePool.Resolution> sharedResolutionCache =
-      CacheBuilder.newBuilder()
-          .softValues()
+  final ConcurrentMap<TypeCacheKey, TypePool.Resolution> sharedResolutionCache =
+      new ConcurrentLinkedHashMap.Builder<TypeCacheKey, TypePool.Resolution>()
+          .maximumWeightedCapacity(TYPE_CAPACITY)
```
[the remark this answers is not in the document]
Comment (randomanderson) [reply dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/bytebuddy/DDCachingPoolStrategy.java:70]: `ConcurrentLinkedHashMap` is a bounded concurrent map with least-recently-used eviction: https://github.com/ben-manes/concurrentlinkedhashmap/blob/master/src/main/java/com/googlecode/concurrentlinkedhashmap/ConcurrentLinkedHashMap.java#L108

[the remark this answers is not in the document]
Comment (tylerbenson) [reply dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/bytebuddy/DDCachingPoolStrategy.java:70]: Ah, ok.  I was getting this confused with `ConcurrentHashMap`.

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/CaffeineWeakCache.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  }
+
+  @Override
+  public V getIfPresent(K key) {
+    return cache.getIfPresent(key);
+  }
+
+  @Override
+  public V computeIfAbsent(K key, Function<? super K, ? extends V> mappingFunction) {
+    // Unable to use cache.get() directly because it relies on java.util.function.Function which is
```
[the remark this answers is not in the document]
Comment (bantonsson) [reply dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/CaffeineWeakCache.java:35]: Agree that this is a realistic assumption. The same is true for the implementation in `CLHMWeakCache`, so maybe there should be a comment in the `WeakCache` interface so it doesn't come as a surprise to users of the cache.

**Review by bantonsson — APPROVED** [review]

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/bytebuddy/DDCachingPoolStrategy.java (hunk 2 above)
Comment (tylerbenson) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/bytebuddy/DDCachingPoolStrategy.java:70]: Worth doing a perf test before/after this change?

**Review by tylerbenson — APPROVED** [review]

**Comment by devinsba** [comment]
Fixes #2073

**Comment by mcculls** [comment]
Hi @randomanderson - I think everything should now be in place to switch Caffeine over to our own scheduler.

I've been testing this in #2109 - if you rebase this PR on latest master and cherry-pick these two commits, CI should pass:
* faa701b8d09a4c70788d1538b9ed0a0399bf4d64 use `AgentTaskScheduler.INSTANCE` as Caffeine's executor for background cleanup
* ddd999139dbd356dc843dc35939b68c064f1e5b8 patch Caffeine's `BoundedLocalCache.PerformCleanupTask`
  (to avoid loading `ForkJoinTask` before instrumentation is ready)

The last commit is the smallest way I found to safely patch the offending task that extends `ForkJoinTask` and causes `ForkJoinTask` to load before instrumentation is ready. An alternative approach would be to compile a complete copy of the Caffeine source in a separate module (with the necessary source patch) and use that module instead of the inlined dependency.

**Review by richardstartin — APPROVED** [review]
I'm ok with the patch

**Review by bantonsson — APPROVED** [review]

**Pull request merged**